### PR TITLE
Add experimental POETRY_REQUESTS_TIMEOUT option

### PR DESCRIPTION
**Pull request by Secrus**
# Pull Request Check List

Resolves: #6009

- [ ] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

Not sure if it should be tested, since it is only a stdlib function usage.

### CODE REVIEW

**Comment by mbbyn** [comment]
> Is there any appetite to merge this change? It would really help anyone trying to download larger packages that are being scanned by corporate proxies. Without being able to extend this timeout, downloading packages like pyspark or pytorch results in a failure.

We faced this issue for a while and eventually became too painful to ignore, so we set up a devpi server internally with docker (first we set up pypiserver, but devpi was better for mirroring pypi.org). You may or may not be able to do this, but do consider it even if deployed locally it has proven to be a worthwhile 2-day investment.

**Comment by ralbertazzi** [comment]
Although I'm not facing this issue, it seems more than reasonable to me that users have a way to configure network timeouts. This PR looks a no brainer to me, we have real users waiting for it, and there's already been consensus on making this a quick temporary solution. But it's of no use if it's not _quick_ :) I would encourage a merge 🙏 @radoering @dimbleby

**Comment by dimbleby** [comment]
(no point summoning me, I do not have the commit bit)

File: docs/faq.md [hunk 1]
```diff
@@ -152,3 +152,16 @@ This is done so to be compliant with the broader Python ecosystem.
...
+
+### My requests are timing out!
+
+Poetry's default HTTP request time out is 15 seconds, the same as `pip`.
+Similar to `PIP_REQUESTS_TIMEOUT`, the experimental environment variable `POETRY_REQUESTS_TIMEOUT`
+can be set to alter this value. As this environment variable is **unstable**, it may be removed in the future, but
+not before being replaced with a permanent alternative.
+
+Note that it is almost always better to debug network stability than to blindly increase
+the request timeout -- the timeout controls how long it Poetry waits to establish a HTTP
```
[the remark this answers is not in the document]
Comment (radoering) [reply docs/faq.md:236]: I agree, we should be more accurate here. I'm not even sure if debugging network stability really helps. I'm not sure about this, but from observation I believe a read timeout can also occur when the server is overloaded. IMO we'd better omit this paragraph entirely if we don't know how to write it correctly.

**Review by radoering — CHANGES_REQUESTED** [review]
> This PR looks a no brainer to me

It does, but there are at least two things to consider:

- Do we really want to introduce a config option that can only be configured via an environment variable and not via `poetry config`? (We have already discussed that we accept that in this case.)
- Documentation should be accurate. (see comments)

File: docs/faq.md [hunk 2]
```diff
@@ -152,3 +152,16 @@ This is done so to be compliant with the broader Python ecosystem.
...
 For example, if Poetry builds a distribution for a project that uses a version that is not valid according to
 [PEP 440](https://peps.python.org/pep-0440), third party tools will be unable to parse the version correctly.
+
+
+### My requests are timing out!
+
+Poetry's default HTTP request time out is 15 seconds, the same as `pip`.
+Similar to `PIP_REQUESTS_TIMEOUT`, the experimental environment variable `POETRY_REQUESTS_TIMEOUT`
+can be set to alter this value. As this environment variable is **unstable**, it may be removed in the future, but
+not before being replaced with a permanent alternative.
```
[the remark this answers is not in the document]
Comment (Secrus) [reply docs/faq.md]: Dropped as suggested

File: docs/faq.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Secrus) [reply docs/faq.md]: Deleted. Didn't know how to rephrase it better (or what to suggest instead).

File: docs/faq.md [hunk 3]
```diff
@@ -223,3 +223,10 @@ RUN poetry install --no-dev
 
 The two key options we are using here are `--no-root` (skips installing the project source) and `--no-directory` (skips installing any local directory path dependencies, you can omit this if you don't have any).
 [More information on the options available for `poetry install`]({{< relref "cli#install" >}}).
+
+
+### My requests are timing out!
+
+Poetry's default HTTP request time out is 15 seconds, the same as `pip`.
```
Comment (radoering) [inline docs/faq.md:230]:
```suggestion
Poetry's default HTTP request timeout is 15 seconds, the same as `pip`.
```

Just noticed in the generated docs.

File: docs/faq.md [hunk 4]
```diff
@@ -223,3 +223,10 @@ RUN poetry install --no-dev
...
+Similar to `PIP_REQUESTS_TIMEOUT`, the experimental environment variable `POETRY_REQUESTS_TIMEOUT`
+can be set to alter this value. This environment variable is **unstable**.
```
Comment (radoering) [inline docs/faq.md:232]:
```suggestion
can be set to alter this value.
```

The previous sentence already says _"experimental environment variable"_. You could emphasize **experimental** if you like. IMO, that's sufficient. The last sentence could also be understood to mean that setting the environment variable does not work reliably.

**Review by radoering — CHANGES_REQUESTED** [review]

File: docs/faq.md (hunk 3 above)
Comment (Secrus) [reply docs/faq.md]: Fixed

File: docs/faq.md (hunk 4 above)
Comment (Secrus) [reply docs/faq.md]: Done

**Review by radoering — APPROVED** [review]

**Pull request merged**